### PR TITLE
Make stitching panoramas reusable after estimating transform once

### LINKED ISSUE #4480: Stitcher is one-time use
Transferred from http://code.opencv.org/issues/2675

```
|| Todd Pepper on 2013-01-02 16:28
|| Priority: Normal
|| Affected: None
|| Category: stitching
|| Tracker: Bug
|| Difficulty: None
|| PR: 
|| Platform: None / None
```
## Stitcher is one-time use

```
It seems that the stitcher module is one-time use as a call to composePanorama will change camera parameters such that subsequent calls will improperly place images.
It would be useful to be able to stitch one set of images and then apply the solved parameters to another set of images.

I think the following code may be part of the problem:

<pre>
Stitcher::Status Stitcher::composePanorama(InputArray images, OutputArray pano)
...

            // Update corners and sizes
            for (size_t i = 0; i < imgs_.size(); ++i)
            {
                // Update intrinsics
                cameras_[i].focal *= compose_work_aspect;
                cameras_[i].ppx *= compose_work_aspect;
                cameras_[i].ppy *= compose_work_aspect;

                // Update corner and size

...

</pre>
```
##### Daniil Osokin on 2013-01-09 12:35

```
Hi, Todd!
It would be nice, if you can modify stitcher and send a pull request to OpenCV.
```
##### Kirill Kornyakov on 2013-01-09 15:17

```
Alexey, could you please comment if this is a bug? Can it be that the code intentionally modifies parameters?
```
##### Alexey Spizhevoy on 2013-01-10 06:21

```
Confirm, this is a bug.
```
##### Todd Pepper on 2013-03-06 15:40

```
Daniil Osokin wrote:
> Hi, Todd!
> It would be nice, if you can modify stitcher and send a pull request to OpenCV.

I will see what I can do.
```
##### Todd Pepper on 2013-03-27 12:51

```
Daniil Osokin wrote:
> Hi, Todd!
> It would be nice, if you can modify stitcher and send a pull request to OpenCV.

Daniil, please point me to instructions on sending a pull request to OpenCV.  

Thank you.
```
##### Sergio Basurco on 2013-10-03 07:53

```
Todd Pepper wrote:
> Daniil Osokin wrote:
> > Hi, Todd!
> > It would be nice, if you can modify stitcher and send a pull request to OpenCV.
> 
> Daniil, please point me to instructions on sending a pull request to OpenCV.  
> 
> Thank you.

Hi, take a look at this link to know about contributing and pull requests [[http://code.opencv.org/projects/opencv/wiki/How_to_contribute?version=38]]
Were you able to fix this issue?
```
##### Todd Pepper on 2013-10-14 13:08

```
Sergio,

No, I have no fix at this time.
```

### PR DESCRIPTION
**Pull request by superbort**
This resolves #4480.

Stitcher will now make a working copy of the CameraParams object to avoid side effects when composing Panorama.
Makes it possible to estimate transform once and then compose multiple panoramas. Useful for setup with fixed cameras.

### CODE REVIEW

**Comment by alalek** [comment]
No need to reopen new PR if you want to add more changes (or remove something from the patch).

Please avoid using of these branch names: `master`/`2.4`. These branches are not scheduled on our Buildbot (there are some issues with validation check)

File: modules/stitching/src/camera.cpp [hunk 1]
```diff
@@ -63,10 +63,12 @@ const CameraParams& CameraParams::operator =(const CameraParams &other)
...
-    Mat_<double> k = Mat::eye(3, 3, CV_64F);
-    k(0,0) = focal; k(0,2) = ppx;
-    k(1,1) = focal * aspect; k(1,2) = ppy;
-    return k;
+  Mat_<double> k = Mat::eye(3,3, CV_64F);
+  k(0,0) = focal;
+  k(0,2) = ppx;
+  k(1,1) = focal  * aspect;
+  k(1,2) = ppy;
+  return k;
```
Comment (alalek) [inline modules/stitching/src/camera.cpp]: Looks like this change is not needed.
Please revert it and squash all commits into one to remove unrelated changes from git history.

Comment (superbort) [inline modules/stitching/src/camera.cpp]: I have moved to a new branch. Do I need to create a new PR or can I change the branch somehow ?

Comment (alalek) [inline modules/stitching/src/camera.cpp]: For new source branch it is needed to open new PR. GitHub allows to change target branch only at this moment.

**Comment by superbort** [comment]
Closed because source branch changed
